### DISCUSSION TITLE
Can I start with C++ modules instead of pre-compiled header?

### DISCUSSION THREAD

**Discussion by rnikander** [microsoft/WindowsAppSDK#3987] [Q&A]
I'm just starting to learn Windows C++ development. When I create a project in Visual Studio, I like to set the C++ version to `/std:c++latest`, and I want to use modules whenever I can. I created a new project using the template: "Blank App, Packaged (WinUI 3 in Desktop)". I see it uses a `pch.h` precompiled header. 

This [page in MS docs](https://learn.microsoft.com/en-us/cpp/build/compare-inclusion-methods?view=msvc-170) claims that modules are an improvement over pre-compiled headers. Is there a way to replace the `pch.h` here and use modules instead? If I just turn off the setting, I think I see the compilation take a lot longer. I'm not sure how one would arrange files and imports to replace (if possible) the pch.h here.

**Comment by DarranRowe** [comment]
Unfortunately, no.
The closest you could currently get is header units, and the Windows headers tend to not be "well behaved". To get to the point of you being able to use modules in a WinUI 3 project, there would have to be modules for Windows and C++/WinRT.

**Reply by rnikander** [reply]
Okay, thanks. Hopefully I can use modules for my own code without having to wait forever for the entire Windows API to be using modules. I'm going to try and see how it goes.

I also thought I might be able to get a compilation speedup like a pre-compiled header by including things in a global fragment and re-exporting them, like...

```c++
module;
#include [windows api stuff]
export module windows;
export ... (re-export stuff from the headers that I want to use...)
```

Haven't tried it yet. 

I got a different answer here: https://learn.microsoft.com/en-us/answers/questions/1435360/should-i-turn-off-pre-compiled-headers-if-im-using

But I don't have a lot of confidence that they are speaking from experience there...

**Comment by jaigak** [comment]
Also, note that officially there is no statement denying you cannot use both precompiled headers and modules. However, from my experience, it does not work properly if you refer to a type defined in a header that's included in PCH from a module.

**Comment by YexuanXiao** [comment]
Hello everyone, I've added C++ module support to C++/WinRT and written a blog post about my journey. With my fork, you can reference the modules you need on demand, just like in C#. I've published it on NuGet with the ID YexuanXiao.CppWinRTPlus. The repository link is https://github.com/YexuanXiao/cppwinrtplus, and the blog post is at https://github.com/YexuanXiao/cppwinrtplus/blob/master/about_module.md. The project is currently in the experimental stage, and I hope you'll enjoy it. Welcome any bug reports, feature suggestions, and other feedback.

**Comment by RDMacLachlan** [comment]
* #6446
